File: lightbulb-dialogs/src/main/java/com/github/RooneyAndShadows/lightbulb/dialogs/picker_dialog_icon/IconPickerDialogBuilder.java
package com.github.rooneyandshadows.lightbulb.dialogs.picker_dialog_icon;

import com.github.rooneyandshadows.lightbulb.dialogs.base.LightBulbDialogBuilder;
import com.github.rooneyandshadows.lightbulb.dialogs.base.LightBulbDialogFragment;
import com.github.rooneyandshadows.lightbulb.dialogs.base.LightBulbPickerDialogFragment;
import com.github.rooneyandshadows.lightbulb.dialogs.picker_dialog_adapter.AdapterPickerDialog;

import androidx.fragment.app.FragmentManager;

public class IconPickerDialogBuilder extends LightBulbDialogBuilder<IconPickerDialog> {
    private LightBulbPickerDialogFragment.SelectionChangedListener<int[]> changedCallback;
    private int[] selection;
    private final IconPickerAdapter adapter;


    public IconPickerDialogBuilder(FragmentManager manager, String dialogTag, IconPickerAdapter adapter) {
        super(manager, dialogTag);
        this.adapter = adapter;
    }

    @Override
    public IconPickerDialogBuilder withTitle(String title) {
        return (IconPickerDialogBuilder) super.withTitle(title);
    }

    @Override
    public IconPickerDialogBuilder withMessage(String message) {
        return (IconPickerDialogBuilder) super.withMessage(message);
    }

    @Override
    public IconPickerDialogBuilder withPositiveButton(LightBulbDialogFragment.DialogButtonConfiguration positiveButtonConfiguration, LightBulbDialogFragment.DialogButtonClickListener onClickListener) {
        return (IconPickerDialogBuilder) super.withPositiveButton(positiveButtonConfiguration, onClickListener);
    }

    @Override
    public IconPickerDialogBuilder withNegativeButton(LightBulbDialogFragment.DialogButtonConfiguration negativeButtonConfiguration, LightBulbDialogFragment.DialogButtonClickListener onClickListener) {
        return (IconPickerDialogBuilder) super.withNegativeButton(negativeButtonConfiguration, onClickListener);
    }

    @Override
    public IconPickerDialogBuilder withOnCancelListener(LightBulbDialogFragment.DialogCancelListener listener) {
        return (IconPickerDialogBuilder) super.withOnCancelListener(listener);
    }

    @Override
    public IconPickerDialogBuilder withOnShowListener(LightBulbDialogFragment.DialogShowListener listener) {
        return (IconPickerDialogBuilder) super.withOnShowListener(listener);
    }

    @Override
    public IconPickerDialogBuilder withOnHideListener(LightBulbDialogFragment.DialogHideListener listener) {
        return (IconPickerDialogBuilder) super.withOnHideListener(listener);
    }

    @Override
    public IconPickerDialogBuilder withCancelOnClickOutsude(boolean closeOnClickOutside) {
        return (IconPickerDialogBuilder) super.withCancelOnClickOutsude(closeOnClickOutside);
    }

    @Override
    public IconPickerDialogBuilder withDialogType(LightBulbDialogFragment.DialogTypes dialogType) {
        return (IconPickerDialogBuilder) super.withDialogType(dialogType);
    }

    @Override
    public IconPickerDialogBuilder withAnimations(LightBulbDialogFragment.DialogAnimationTypes animation) {
        return (IconPickerDialogBuilder) super.withAnimations(animation);
    }

    public IconPickerDialogBuilder withSelectionCallback(AdapterPickerDialog.SelectionChangedListener<int[]> listener) {
        changedCallback = listener;
        return this;
    }

    public IconPickerDialogBuilder withSelection(int[] selection) {
        this.selection = selection;
        return this;
    }

    @Override
    public IconPickerDialog buildDialog() {
        IconPickerDialog iconPickerDialog = (IconPickerDialog) fragmentManager.findFragmentByTag(dialogTag);
        if (iconPickerDialog == null)
            iconPickerDialog = IconPickerDialog.newInstance(title, message, positiveButtonConfiguration, negativeButtonConfiguration, cancelableOnClickOutside, animation);
        iconPickerDialog.setFragmentManager(fragmentManager);
        iconPickerDialog.setDialogTag(dialogTag);
        iconPickerDialog.addOnShowListener(onShowListener);
        iconPickerDialog.addOnHideListener(onHideListener);
        iconPickerDialog.addOnCancelListener(onCancelListener);
        iconPickerDialog.addOnNegativeClickListeners(onNegativeClickListener);
        iconPickerDialog.addOnPositiveClickListener(onPositiveClickListener);
        iconPickerDialog.setAdapter(adapter);
        iconPickerDialog.setOnSelectionChangedListener(changedCallback);
        iconPickerDialog.setSelection(selection);
        return iconPickerDialog;
    }
}

File: lightbulb-dialogs/src/main/java/com/github/RooneyAndShadows/lightbulb/dialogs/picker_dialog_icon/IconPickerDialogBuilder.kt
package com.github.rooneyandshadows.lightbulb.dialogs.picker_dialog_icon

import androidx.lifecycle.LifecycleOwner
import com.github.rooneyandshadows.lightbulb.dialogs.base.BaseDialogBuilder
import com.github.rooneyandshadows.lightbulb.dialogs.base.BasePickerDialogFragment.SelectionChangedListener
import androidx.fragment.app.FragmentManager
import com.github.rooneyandshadows.lightbulb.dialogs.base.internal.*
import com.github.rooneyandshadows.lightbulb.dialogs.base.internal.callbacks.*
import com.github.rooneyandshadows.lightbulb.dialogs.picker_dialog_datetime.DateTimePickerDialog

@Suppress("unused")
class IconPickerDialogBuilder @JvmOverloads constructor(
    lifecycleOwner: LifecycleOwner? = null,
    manager: FragmentManager,
    dialogTag: String,
    private val adapter: IconPickerAdapter,
) : BaseDialogBuilder<IconPickerDialog>(lifecycleOwner, manager, dialogTag) {
    private var changedCallback: SelectionChangedListener<IntArray?>? = null
    private var selection: IntArray? = null

    @Override
    override fun withTitle(title: String): IconPickerDialogBuilder {
        return super.withTitle(title) as IconPickerDialogBuilder
    }

    @Override
    override fun withMessage(message: String): IconPickerDialogBuilder {
        return super.withMessage(message) as IconPickerDialogBuilder
    }

    @Override
    override fun withPositiveButton(
        configuration: DialogButtonConfiguration,
        onClickListener: DialogButtonClickListener,
    ): IconPickerDialogBuilder {
        return super.withPositiveButton(configuration, onClickListener) as IconPickerDialogBuilder
    }

    @Override
    override fun withNegativeButton(
        configuration: DialogButtonConfiguration,
        onClickListener: DialogButtonClickListener,
    ): IconPickerDialogBuilder {
        return super.withNegativeButton(configuration, onClickListener) as IconPickerDialogBuilder
    }

    @Override
    override fun withOnCancelListener(listener: DialogCancelListener): IconPickerDialogBuilder {
        return super.withOnCancelListener(listener) as IconPickerDialogBuilder
    }

    @Override
    override fun withOnShowListener(listener: DialogShowListener): IconPickerDialogBuilder {
        return super.withOnShowListener(listener) as IconPickerDialogBuilder
    }

    @Override
    override fun withOnHideListener(listener: DialogHideListener): IconPickerDialogBuilder {
        return super.withOnHideListener(listener) as IconPickerDialogBuilder
    }

    @Override
    override fun withCancelOnClickOutside(closeOnClickOutside: Boolean): IconPickerDialogBuilder {
        return super.withCancelOnClickOutside(closeOnClickOutside) as IconPickerDialogBuilder
    }

    @Override
    override fun withDialogType(dialogType: DialogTypes): IconPickerDialogBuilder {
        return super.withDialogType(dialogType) as IconPickerDialogBuilder
    }

    @Override
    override fun withAnimations(animation: DialogAnimationTypes): IconPickerDialogBuilder {
        return super.withAnimations(animation) as IconPickerDialogBuilder
    }

    @Override
    override fun withDialogListeners(listeners: DialogListeners): IconPickerDialogBuilder {
        return super.withDialogListeners(listeners) as IconPickerDialogBuilder
    }

    fun withSelectionCallback(listener: SelectionChangedListener<IntArray?>): IconPickerDialogBuilder {
        changedCallback = listener
        return this
    }

    fun withSelection(selection: IntArray): IconPickerDialogBuilder {
        this.selection = selection
        return this
    }

    @Override
    override fun buildDialog(): IconPickerDialog {
        return getExistingDialogOrCreate().apply {
            setLifecycleOwner(dialogLifecycleOwner)
            setDialogCallbacks(dialogListeners)
            setParentFragManager(dialogParentFragmentManager)
            setDialogTag(dialogTag)
            onShowListener?.apply { addOnShowListener(this) }
            onHideListener?.apply { addOnHideListener(this) }
            onCancelListener?.apply { addOnCancelListener(this) }
            onNegativeClickListener?.apply { addOnNegativeClickListeners(this) }
            onPositiveClickListener?.apply { addOnPositiveClickListener(this) }
            setAdapter(adapter)
            setOnSelectionChangedListener(changedCallback)
            setSelection(selection)
        }
    }

    private fun getExistingDialogOrCreate(): IconPickerDialog {
        val dialog = dialogParentFragmentManager.findFragmentByTag(dialogTag) as IconPickerDialog?
        return dialog ?: IconPickerDialog.newInstance().apply {
            dialogTitle = title
            dialogMessage = message
            dialogPositiveButton = positiveButtonConfiguration
            dialogNegativeButton = negativeButtonConfiguration
            isCancelable = cancelableOnClickOutside
            dialogAnimationType = animation
        }
    }
}